File: 2020/day/11/src/Main.hs
{-# LANGUAGE ViewPatterns #-}
{-# OPTIONS_GHC -Wall -Wextra -Werror -Wno-name-shadowing #-}

module Main where

import Control.Monad (filterM, mapM_)
import Control.Monad.State.Strict (State, execState, gets, modify)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Set (Set)
import qualified Data.Set as Set
import System.Environment (getArgs)

main :: IO ()
main = (getArgs >>=) . mapM_ $ \filename -> do
  putStrLn filename

  grid <- parseGrid <$> readFile filename

  putStrLn "\tpart 1: number of filled seats in stable seating configuration"
  putStrLn $ "\t" ++ show (part1 grid)

  putStrLn "\tpart 2: number of filled seats in revised stable seating configuration"
  putStrLn $ "\t" ++ show (part2 grid)

type Grid = Map Position Tile

type Position = (Int, Int)

data Tile = Floor | Unstable | Empty | Full
  deriving (Eq)

parseGrid :: String -> Grid
parseGrid src = Map.fromList $ do
  (y, line) <- zip [0 ..] $ lines src
  (x, char) <- zip [0 ..] line
  return
    ( (x, y),
      case char of
        '.' -> Floor
        'L' -> Unstable
        _ -> error $ "unexpected seating layout symbol " ++ show char ++ ", expected 'L' or '.'"
    )

part1 :: Grid -> Int
part1 = countFull . stabilize adjacent 4

part2 :: Grid -> Int
part2 = countFull . stabilize visible 5

countFull :: Grid -> Int
countFull = length . filter (== Full) . Map.elems

stabilize :: (Position -> Grid -> Grid) -> Int -> Grid -> Grid
stabilize neighbors threshold = execState . loop =<< id
  where
    loop :: Grid -> State Grid ()
    loop (Map.keys . Map.filter (== Unstable) -> candidates) = do
      isFillableNow <- gets (flip isFillable)
      case filterM isFillableNow candidates of
        (_, []) -> return ()
        (Set.toList -> flagged, fillable) -> do
          setAll fillable Full
          unfillable <- filterM (gets . isUnstable) flagged
          setAll unfillable Empty
          loop . Map.unions =<< traverse (gets . neighbors) unfillable

    setAll :: [Position] -> Tile -> State Grid ()
    setAll ps tile = modify . Map.union $ Map.fromList [(p, tile) | p <- ps]

    isFillable :: Position -> Grid -> (Set Position, Bool)
    isFillable p grid = do
      let chairs = Map.keysSet . Map.filter (`elem` [Unstable, Full]) $ neighbors p grid
      if Set.size chairs >= threshold
        then (Set.empty, False)
        else (chairs, True)

    isUnstable :: Position -> Grid -> Bool
    isUnstable p grid = Just Unstable == Map.lookup p grid

adjacent :: Position -> Grid -> Grid
adjacent (x, y) grid = Map.fromList $ do
  (dx, dy) <- deltas
  let p = (x + dx, y + dy)
  case Map.lookup p grid of
    Nothing -> []
    Just tile -> return (p, tile)

visible :: Position -> Grid -> Grid
visible (x, y) grid = Map.fromList $ do
  (dx, dy) <- deltas
  let loop p@(x, y) = case Map.lookup p grid of
        Nothing -> []
        Just Floor -> loop (x + dx, y + dy)
        Just tile -> return (p, tile)
  loop (x + dx, y + dy)

deltas :: [Position]
deltas =
  [ (-1, -1),
    (-1, 0),
    (-1, 1),
    (0, -1),
    (0, 1),
    (1, -1),
    (1, 0),
    (1, 1)
  ]

File: 2020/day/11/src/main.rs
use std::error;

type Error = Box<dyn error::Error>;

fn main() -> Result<(), Error> {
    for filename in std::env::args().skip(1) {
        println!("{}", filename);

        let grid = parse_grid(&std::fs::read_to_string(filename)?)?;

        println!("\tpart 1: number of filled seats in stable seating configuration");
        println!("\t{}", part1(grid.clone()));

        println!("\tpart 2: number of filled seats in revised stable seating configuration");
        println!("\t{}", part2(grid));
    }

    Ok(())
}

type Grid = Vec<Vec<Tile>>;
type Point = (usize, usize);

#[derive(Copy, Clone, PartialEq, Eq, Debug)]
enum Tile {
    Floor,
    Unstable,
    Full,
    Empty,
}
use Tile::*;

fn count_full(grid: Grid) -> usize {
    let mut num_full = 0;
    for row in grid {
        for tile in row {
            if Full == tile {
                num_full += 1;
            }
        }
    }
    num_full
}

fn stabilize(
    mut grid: Grid,
    neighbors: fn(&Grid, Point) -> Vec<(Point, Tile)>,
    threshold: usize,
) -> Grid {
    use std::collections::HashSet;

    let mut queue = HashSet::new();
    for (y, row) in grid.iter().enumerate() {
        for (x, &tile) in row.iter().enumerate() {
            if tile == Unstable {
                queue.insert((x, y));
            }
        }
    }

    while !queue.is_empty() {
        let mut fillable = vec![];
        let mut flagged = HashSet::new();

        for (x, y) in queue.into_iter() {
            let mut num_full = 0;
            let mut unstable = vec![];

            for (p, tile) in neighbors(&grid, (x, y)) {
                match tile {
                    Full => num_full += 1,
                    Unstable => unstable.push(p),
                    _ => (),
                }
            }

            if num_full + unstable.len() < threshold {
                fillable.push((x, y));
                flagged.extend(&unstable);
            }
        }

        for (x, y) in fillable.into_iter() {
            grid[y][x] = Full;
        }

        let mut unfillable = vec![];
        for (x, y) in flagged.into_iter() {
            if grid[y][x] == Full {
                continue;
            }
            grid[y][x] = Empty;
            unfillable.push((x, y));
        }

        queue = HashSet::new();
        for (x, y) in unfillable.into_iter() {
            for (p, tile) in neighbors(&grid, (x, y)) {
                if tile == Unstable {
                    queue.insert(p);
                }
            }
        }
    }

    grid
}

#[allow(unused)]
fn slow_stabilize(
    mut grid: Grid,
    neighbors: fn(&Grid, Point) -> Vec<(Point, Tile)>,
    threshold: usize,
) -> Grid {
    let height = grid.len();
    let width = grid[0].len();
    let mut unstable = true;

    while unstable {
        let mut updates = vec![];

        for y in 0..height {
            for x in 0..width {
                let tile = grid[y][x];

                if tile == Floor {
                    continue;
                }

                let mut num_full = 0;
                for (_p, tile) in neighbors(&grid, (x, y)) {
                    if tile == Full {
                        num_full += 1;
                    }
                }

                if num_full >= threshold && tile != Empty {
                    updates.push(((x, y), Empty));
                } else if num_full == 0 && tile != Full {
                    updates.push(((x, y), Full));
                }
            }
        }

        unstable = updates.len() > 0;

        for ((x, y), tile) in updates.into_iter() {
            grid[y][x] = tile;
        }
    }

    grid
}

const DELTAS: [(i64, i64); 8] = [
    (-1, -1),
    (-1, 0),
    (-1, 1),
    (0, -1),
    (0, 1),
    (1, -1),
    (1, 0),
    (1, 1),
];

fn adjacent(grid: &Grid, (x, y): Point) -> Vec<(Point, Tile)> {
    let height = grid.len() as i64;
    let width = grid[0].len() as i64;

    DELTAS
        .iter()
        .filter_map(|(dx, dy)| {
            let x = x as i64 + dx;
            let y = y as i64 + dy;
            if (0..height).contains(&y) && (0..width).contains(&x) {
                let x = x as usize;
                let y = y as usize;
                Some(((x, y), grid[y][x]))
            } else {
                None
            }
        })
        .collect()
}

fn visible(grid: &Grid, (x, y): Point) -> Vec<(Point, Tile)> {
    let height = grid.len() as i64;
    let width = grid[0].len() as i64;

    DELTAS
        .iter()
        .filter_map(|(dx, dy)| {
            let mut x = x as i64 + dx;
            let mut y = y as i64 + dy;

            while (0..height).contains(&y)
                && (0..width).contains(&x)
                && grid[y as usize][x as usize] == Floor
            {
                x += dx;
                y += dy;
            }

            if (0..height).contains(&y) && (0..width).contains(&x) {
                let x = x as usize;
                let y = y as usize;
                Some(((x, y), grid[y][x]))
            } else {
                None
            }
        })
        .collect()
}

fn part1(grid: Grid) -> usize {
    count_full(stabilize(grid, adjacent, 4))
}

fn part2(grid: Grid) -> usize {
    count_full(stabilize(grid, visible, 5))
}

fn parse_grid(src: &str) -> Result<Grid, Error> {
    src.lines().map(parse_tiles).collect()
}

fn parse_tiles(src: &str) -> Result<Vec<Tile>, Error> {
    src.chars().map(parse_tile).collect()
}

macro_rules! error {
    ($fmt:literal $(, $arg:expr)*) => { Error::from(format!($fmt $(, $arg)*)) }
}

fn parse_tile(src: char) -> Result<Tile, Error> {
    match src {
        'L' => Ok(Unstable),
        '.' => Ok(Floor),
        _ => Err(error!(
            "unrecognized grid character {:?}, expected 'L' or '.'",
            src
        )),
    }
}
